Upgrade to repo-config v0.7.5 with regenerated templates

noah-kreutzer-frequenz opened the pull request:
with manual removals according to upgrades from v0.6.2 to v0.7.5
s/o to cwasicki

cwasicki committed:
Clean up cookiecutter-replay

Signed-off-by: cwasicki <126617870+cwasicki@users.noreply.github.com>

cwasicki committed:
Upgrade to repo-config v0.7.5 with regenerated templates

Manual removal of some files.

Signed-off-by: cwasicki <126617870+cwasicki@users.noreply.github.com>

shsms commented:
@llucax could you also please check if @noah-kreutzer-frequenz is missing some permissions?  Because CI doesn't automatically run, but requires approval.

llucax commented on .github/ISSUE_TEMPLATE/bug.yml:
```diff
@@ -50,6 +50,9 @@ body:
         - Documentation (part:docs)
         - Unit, integration and performance tests (part:tests)
         - Build script, CI, dependencies, etc. (part:tooling)
+        # TODO(cookiecutter): Add other parts
+        # Please have in mind that that the part:xxx labels need to
+        # be created in the GitHub repository.
```
General comments about all TODOs added here: You should probably address them and remove them. Since this is a preexisting repository, it is very likely you want to remove them without much thought, as the setup should have been already done.

We should have a way to determine if this is a first time generation or not and only put `TODOs` for first time generation, but for now it needs to be manually cleaned up on every update :disappointed: 

* https://github.com/frequenz-floss/frequenz-repo-config-python/issues/186

cwasicki replied:
I recommended to leave them in since they will come back anyway :) But I realize they could cause problems with nox, right?

llucax replied:
For the python stuff yes, pylint will complain about TODOs

llucax replied:
I'm not sure if you are leaving some TODOs intentionally, I've seen that you removed some, but some other are still here.

llucax commented on .github/containers/nox-cross-arch/arm64-ubuntu-20.04-python-3.11.Dockerfile:
We are [still deciding](https://github.com/frequenz-floss/frequenz-repo-config-python/discussions/177) if we want to test for arm64, and how, and in which type of repo. For now I would recommend removing all the files related to arm64 testing. I will comment on everything that needs removing regarding arm64.

We should also have a repo-config option to avoid generating arm64 testing to avoid having to clean up all this over and over again when re-generating templates:

* https://github.com/frequenz-floss/frequenz-repo-config-python/issues/187

noah-kreutzer-frequenz replied:
does this mean this whole file should be deleted? @llucax

llucax replied:
Yes

llucax commented on .github/containers/nox-cross-arch/entrypoint.bash:
This file is amr64-related, it should be removed.

noah-kreutzer-frequenz replied:
done.

llucax commented on .github/workflows/ci.yaml:
```diff
@@ -102,12 +103,137 @@ jobs:
+      - name: Set pip cache owners to root for docker
+        run: if [[ -e /tmp/pip-cache ]]; then sudo chown -R root:root /tmp/pip-cache; fi
+
+      - name: Run nox
+        run: |
+          docker run \
+            --rm \
+            -v $(pwd):/${{ github.workspace }} \
+            -v /tmp/pip-cache:/root/.cache/pip \
+            -w ${{ github.workspace }} \
+            --net=host \
+            --platform linux/${{ matrix.arch }} \
+            localhost/nox-cross-arch:latest \
+            bash -c "pip install -e .[dev-noxfile]; nox --install-only -e ${{ matrix.nox-session }}; pip freeze; nox -e ${{ matrix.nox-session }}"
+        timeout-minutes: 30
+
+      # This ensures that the runner has access to the pip cache.
+      - name: Reset pip cache ownership
+        if: always()
+        run: sudo chown -R $USER:$USER /tmp/pip-cache
+
+  # This job runs if all the `nox-cross-arch` matrix jobs ran and succeeded.
+  # As the `nox-all` job, its main purpose is to provide a single point of
+  # reference in branch protection rules, similar to how `nox-all` operates.
+  # However, there's a crucial difference: the `nox-cross-arch` job is omitted
+  # in PRs. Without the `nox-cross-arch-all` job, the inner matrix wouldn't be
+  # expanded in such scenarios. This would lead to the CI indefinitely waiting
+  # for these jobs to complete due to the branch protection rules, essentially
+  # causing it to hang. This behavior is tied to a recognized GitHub matrices
+  # issue when certain jobs are skipped. For a deeper understanding, refer to:
+  # https://github.com/orgs/community/discussions/9141
+  nox-cross-arch-all:
+    # The job name should match the name of the `nox-cross-arch` job.
+    name: Cross-arch tests with nox
+    needs: ["nox-cross-arch"]
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Return true
+        run: "true"
+
```
This is amr64-related, it should be removed.

```suggestion
```

llucax commented on .github/workflows/ci.yaml:
```diff
@@ -174,84 +325,74 @@ jobs:
   publish-docs:
     name: Publish documentation website to GitHub pages
-    needs: ["nox", "build", "protolint"]
+    needs: ["nox-all", "nox-cross-arch-all", "test-installation", "protolint"]
```
This is amr64-related, it should be changed.
```suggestion
    needs: ["nox-all", "test-installation", "protolint"]
```

llucax commented on CONTRIBUTING.md:
````diff
@@ -198,3 +198,30 @@ These are the steps to create a new release:
    eventually too).
 
 7. Celebrate!
+
+##  Cross-Arch Testing
+
+This project has built-in support for testing across multiple architectures.
+Currently, our CI conducts tests on `arm64` machines using QEMU emulation. We
+also have the flexibility to expand this support to include additional
+architectures in the future.
+
+This project contains Dockerfiles that can be used in the CI to test the
+python package in non-native machine architectures, e.g., `arm64`. The
+Dockerfiles exist in the directory `.github/containers/nox-cross-arch`, and
+follow a naming scheme so that they can be easily used in build matrices in the
+CI, in `nox-cross-arch` job. The naming scheme is:
+
+```
+<arch>-<os>-python-<python-version>.Dockerfile
+```
+
+E.g.,
+
+```
+arm64-ubuntu-20.04-python-3.11.Dockerfile
+```
+
+If a Dockerfile for your desired target architecture, OS, and python version
+does not exist here, please add one before proceeding to add your options to
+the test matrix.
````
This is amr64-related, it should be removed.
```suggestion
```

llucax commented on .gitignore:
```diff
@@ -3,12 +3,6 @@ __pycache__/
 *.py[cod]
 *$py.class
 
-# Auto-generated python files from the protocol buffer compiler
-py/**/*_pb2.py
-py/**/*_pb2.pyi
-py/**/*_pb2_grpc.py
-py/**/*_pb2_grpc.pyi
-
```
Mmm, interesting that this got removed, I think it should stay, because these files will be generated as soon as you use `pip install -e .`, and you don't want to add them to the repo accidentally. We should add this too to repo-config:

* https://github.com/frequenz-floss/frequenz-repo-config-python/issues/185

llucax commented on .github/workflows/ci.yaml:
```diff
@@ -133,13 +259,38 @@ jobs:
           path: dist/
           if-no-files-found: error
 
+  test-installation:
+    name: Test package installation in different architectures
+    needs: ["build"]
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Fetch sources
+        uses: actions/checkout@v4
+      - name: Download package
+        uses: actions/download-artifact@v3
+        with:
+          name: dist-packages
+          path: dist
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v3
+      - name: Set up docker-buildx
+        uses: docker/setup-buildx-action@v3
+      - name: Test Installation
+        uses: docker/build-push-action@v5
+        with:
+          context: .
+          file: .github/containers/test-installation/Dockerfile
+          platforms: linux/amd64,linux/arm64
+          tags: localhost/test-installation
+          push: false
```
Just FYI, this is arm64-realted, but it only tests that the package can be installed, is more lightweight than the full testing in arm64 and it runs early in the workflow and as a parallel step, so it usually doesn't affect the total workflow run time, so I would leave it.

llucax commented on README.md:
```diff
@@ -4,49 +4,21 @@
-## Overview
-This Weather Forecast API provides access to weather forecast data. The forecast 
-data includes features such as wind components and solar radiation levels at
-specified geographic coordinates. The forecasting process involves predicting
-atmospheric conditions based on various factors and models. The data provided by
-this API represents the forecasted state of the atmosphere at different heights
-above the ground and for various forecast creation times. 
-
-## Objective
-The primary objective of this API is to provide real-time and historical weather
-forecast data to enable more efficient
-energy management, trading, and planning within microgrids and larger gridpool
-portfolios.
-
-## Key Features
-- Real-time Weather Forecasting: Access current atmospheric conditions based on
- recent data.
-- Historical Data Access: Retrieve past weather forecasts for analysis and trend
- prediction.
-- Feature Customization: Choose from a variety of weather features like wind speed,
- solar radiation, and more.
-- Geographical Flexibility: Get forecasts for any location specified by latitude and
- longitude.
-
-## Example Use Cases
-- Adjusting the trading of photovoltaic (PV) surplus production based on future
- solar radiation forecasts.
-- Feeding excess energy into batteries when low consumption or high production is
- forecasted.
-- Dynamic pricing strategies for energy trading based on anticipated weather
- conditions.
-- Resource allocation in microgrids for anticipated energy demands, guided by
- weather forecasts.
-- Risk assessment and contingency planning for severe weather events.
-
-## Target Audience
-This API is designed for application developers in the energy sector who focus on
-the tasks of optimizing microgrid electricity flows or trading of energy. Its design
-aims to be as developer-friendly as possible, requiring no prior knowledge in
-electrical engineering and systems and/or weather forecast services. 
```
All of this **should not** be removed!

llucax commented on RELEASE_NOTES.md:
```diff
@@ -2,13 +2,16 @@
+<!-- Here goes notes on how to upgrade from previous versions, including deprecations and what they should be replaced with -->
 
 ## New Features
 
-- Introduce Pagination to the historical RPC of the service.
+<!-- Here goes the main new features and examples or instructions on how to use them -->
```
This probably should not be removed either.

llucax commented on noxfile.py:
```diff
@@ -1,5 +1,5 @@
 # License: MIT
-# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
```
This should not be updated.
```suggestion
# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
```

llucax commented on proto/frequenz/api/weather/weather.proto:
This file also **should not** be updated!

llucax commented on py/frequenz/api/weather/__init__.py:
This file also **should not** be updated!

llucax commented on pyproject.toml:
```diff
@@ -5,7 +5,7 @@
   "setuptools == 68.1.0",
   "setuptools_scm[toml] == 7.1.0",
-  "frequenz-repo-config[api] == 0.6.2",
+  "frequenz-repo-config[api] == 0.7.0",
```
There is a but in repo-config v0.7.5, it is using an older release, this should be 0.7.5.
```suggestion
  "frequenz-repo-config[api] == 0.7.5",
```

llucax commented on pyproject.toml:
```diff
@@ -25,8 +26,9 @@ classifiers = [
+# TODO(cookiecutter): Remove and add more dependencies if appropriate
 dependencies = [
-  "frequenz-api-common >= 0.5.0, < 0.6.0",
+  "frequenz-api-common >= 0.3.1, < 0.4.0",
```
You probably don't want to downgrade this.
```suggestion
  "frequenz-api-common >= 0.5.0, < 0.6.0",
```

llucax commented on pyproject.toml:
```diff
@@ -36,50 +38,53 @@ dynamic = ["version"]
-  "frequenz-repo-config[api] == 0.6.2",
+  "mkdocs-literate-nav == 0.6.1",
+  "mkdocs-macros-plugin == 1.0.4",
+  "mkdocs-material == 9.3.1",
```
You probably don't want to downgrade `mkdocs-material`, and actually you probably don't want to downgrade **any** dependency.

llucax commented on pyproject.toml:
```diff
@@ -36,50 +38,53 @@ dynamic = ["version"]
+  "mkdocs-macros-plugin == 1.0.4",
+  "mkdocs-material == 9.3.1",
+  "mkdocstrings[python] == 0.23.0",
+  "frequenz-repo-config[api] == 0.7.0",
```
```suggestion
  "frequenz-repo-config[api] == 0.7.5",
```

llucax replied:
This one still needs to be updated

llucax commented on pyproject.toml:
```diff
@@ -36,50 +38,53 @@ dynamic = ["version"]
 dev-noxfile = [
   "nox == 2023.4.22",
-  "frequenz-repo-config[api] == 0.6.2",
+  "frequenz-repo-config[api] == 0.7.0",
```
```suggestion
  "frequenz-repo-config[api] == 0.7.5",
```

llucax commented on pyproject.toml:
```diff
@@ -36,50 +38,53 @@ dynamic = ["version"]
-  "pytest == 7.4.0",
-  "frequenz-repo-config[extra-lint-examples] == 0.6.2",
+  "pytest == 7.4.2",
+  "frequenz-repo-config[extra-lint-examples] == 0.7.0",
```
```suggestion
  "frequenz-repo-config[extra-lint-examples] == 0.7.5",
```

llucax commented on mkdocs.yml:
```diff
@@ -3,29 +3,33 @@
     edit: material/file-edit-outline
     repo: fontawesome/brands/github
-  custom_dir: docs/overrides
+  custom_dir: docs/_overrides
```
You need to remove the old directory `docs/overrides` too, please have a look at the [updating notes in the releases notes](https://github.com/frequenz-floss/frequenz-repo-config-python/releases) (for all the versions you are jumping too).

llucax commented on mkdocs.yml:
```diff
@@ -3,29 +3,33 @@
 # Build directories
 theme:
   name: "material"
-  logo: logo.png
```
You also need to remove the old logo file in `docs/logo.png`.

llucax commented on mkdocs.yml:
```diff
@@ -57,13 +62,15 @@ extra:
     default: latest
 
 extra_css:
-  - css/style.css
-  - css/mkdocstrings.css
```
Same here, you need to remove the old directory `docs/css` and its files.

llucax reviewed (commented):
I'm really sorry but at the moment updating the templates in repo-config need a lot of manual work. There is a lot that needs to be changed (or not changed) :disappointed:

noah-kreutzer-frequenz committed:
Finalize repository configuration updates v0.7.5

This commit consolidates multiple changes made during the repository configuration update. The following adjustments have been made:

Remove TODO in .github/ISSUE_TEMPLATE/config.yml.
Remove .bash file.
Remove TODOs from configuration files.
Remove TODOs and update versions in pyproject.toml.
Revert changes to older versions in various files.
Delete old files and directories in the docs folder for v0.7.0 update.
Remove arm64-related configurations.
Edit CODEOWNERS.
Remove remaining TODO in mkdocs.yml.

These changes ensure a clean repository configuration state, updates to documentation, and version alignment with v0.7.5.

Signed-off-by: noah-kreutzer-frequenz <noah.kreutzer@frequenz.com>

llucax commented on CONTRIBUTING.md:
Not sure why you are adding and removing empty lines in this file...

noah-kreutzer-frequenz replied:
oh, I actually didin't do it on purpose. Don't know what happend here, sorry

llucax commented on mkdocs.yml:
```diff
@@ -3,29 +3,33 @@
 repo_name: "frequenz-api-weather"
 repo_url: "https://github.com/frequenz-floss/frequenz-api-weather"
+# TODO(cookiecutter): "main" is the GitHub repo default branch, you might want to update it
+# if the project uses a different default branch.
```
In this file there are also a lot of TODOs left.

llucax reviewed (commented):
Just a few things missing.

Also, the commit messages are not super useful, not sure if you plan to squash most of it together before merging. If you haven't yet, I recommend having a look at: https://github.com/frequenz-io/rnd/blob/master/doc/coding-standards/commit-messages.md

llucax commented on CONTRIBUTING.md:
```diff
@@ -197,4 +197,4 @@ These are the steps to create a new release:
    eventually too).
 
-7. Celebrate!
+7. Celebrate!
```
Weird, I can't see what changed here. :thinking:

llucax commented:
Since there are no visible changes to the user, you can add the label `cmd:skip-release-notes` to make the check `Release Notes Check` pass, I now added the label myself.

llucax commented:
Since there are no visible changes to the user, you can add the label `cmd:skip-release-notes` to make the check `Release Notes Check` pass, I now added the label myself.

noah-kreutzer-frequenz commented:
I have merged my commit into one I hope, it's fine that way.
Since this is my first official PR I would be very grateful to recieve some further feedback @llucax @shsms

llucax reviewed (approved):
LGTM, only 2 minor comments:

1. Commit messages should be wrapped at 80 columns
2. I would have squashed the last 2 commits together, just having one commit to undo half of it in the following commit makes reviewing much harder

noah-kreutzer-frequenz commented:
Alright I'll have that in mind for the next time.
Thanks for your tipps and patience @shsms @llucax @cwasicki 🙏🏾
